Address issues raised by recent reviews

aphillips opened the pull request:
_editor's copy locations include the fragment reference and are linked to @aphillips's repo version, since that displays more nicely than the diff/preview)_

Addresses #88
Add guidance for generating bidirectional labels/descriptions
Caused by w3c/i18n-activity#1638
[editor's copy location #strcat](https://aphillips.github.io/bp-i18n-specdev/#strcat)

Addresses #83
Add guidance on Unicode code point order
[editor's copy location #char_sort](https://aphillips.github.io/bp-i18n-specdev/#char_sort)

Addresses #69
Add information about constituent parts of BCP47 (specifically 4647)
[editor's copy location #lang_bcp_not_rfc](https://aphillips.github.io/bp-i18n-specdev/#lang_bcp_not_rfc)
[editor's copy location #lang_matching_bcp](https://aphillips.github.io/bp-i18n-specdev/#lang_matching_bcp)

Addresses #14 
Add guidance on using the ILSTR instead of making your own list of subtags
[editor's copy location #use_lstr](https://aphillips.github.io/bp-i18n-specdev/#use_lstr)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/aphillips/bp-i18n-specdev/pull/89.html" title="Last updated on Mar 16, 2023, 5:23 PM UTC (6ce0aea)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/bp-i18n-specdev/89/3e53a99...aphillips:6ce0aea.html" title="Last updated on Mar 16, 2023, 5:23 PM UTC (6ce0aea)">Diff</a>

aphillips committed:
Add guidance for generating bidirectional text from substrings or for
labels/descriptions.
Addresses #88

aphillips committed:
Address #83
Add guidance on Unicode code point order.

aphillips committed:
Address #69
Add a reference to 4647 when talking about BCP47

Also moved requirements above explanation per our standard (thanks
@r12a)

aphillips committed:
Corrections and minor cleanup for the bidi example.

aphillips committed:
Addresses #14
Mention of using the language subtag registry instead of making your own
list.

Also some cleanup of the language tagging of the Arabic examples.

r12a commented on index.html:
```diff
@@ -351,7 +351,15 @@ <h3>Defining language values</h3>
+		<p class="advisement">Specifications SHOULD refer to the IANA Language Subtag Registry instead of providing lists of codes extracted from ISO 639, ISO 3166, or other standards.</p>
+ 	</div>
+ 	
+ 	<p>As part of BCP 47, IANA maintains the language subtag registry, which is a publicly available, machine-readable list of valid subtags for use in language tags. While this registry is based on underlying ISO standards, such as ISO 639 (languages) and ISO 3166 (regions), the list is actively maintained, stabilized, and comprehensive in ways that other lists found on the Internet may not be. Ea ...
```
You could also add something like:
"Selecting one of the ISO or UN code lists will unnecessarily restrict the list of languages that can be described."

aphillips replied:
Good callout. I did a more extensive edit, plus I added links to the "choosing a language tag" article, your subtag lookup tool, and the registry itself.

r12a commented on index.html:
```diff
@@ -951,15 +959,15 @@ <h3>Generating or requiring creation of mixed direction strings</h3>
 &#x0645;&#x0627;&#x0631;&#x0643;&#x0629; A123B (1920 x 1080) 36" &#x0634;&#x0627;&#x0634;&#x0629; &#x0627;&#x0644;&#x0643;&#x0645;&#x0628;&#x064A;&#x0648;&#x062A;&#x0631;, 75 Hz, 4 &#x0645;&#x0644;&#x0644;&#x064A; &#x062B;&#x0627;&#x0646;&#x064A;&#x0629;, &#x0645;&#x0643;&#x0628;&#x0631;&#x0627;&#x062A; &#x0635;&#x0648;&#x062A; &#x0645;&#x062F;&#x0645;&#x062C;&#x0629;</pre>
 
-<p>Notice how the part number (<kbd>A123B</kbd>) is separated from the brand name (<kbd>&#x0645;&#x0627;&#x0631;&#x0643;&#x0629;</kbd>), the measurement <kbd>36</kbd> and the marker for inches (<kbd>"</kbd>) have become separated and that the values <kbd>75 Hz</kbd> (where the measurement is in ASCII) and <kbd>4 ms</kbd> (where the measurement has an Arabic translation <kbd>&#x0645;&#x0644;&#x064 ...
+<p>Notice how the part number (<kbd lang="zxx" translate="no">A123B</kbd>) is separated from the brand name (<kbd lang="ar" dir="rtl" translate="no">&#x0645;&#x0627;&#x0631;&#x0643;&#x0629;</kbd>), the measurement <kbd>36</kbd> and the marker for inches (<kbd>"</kbd>) have become separated and that the values <kbd>75 Hz</kbd> (where the measurement is in ASCII) and <kbd>4 ms</kbd> (where the meas ...
```
If "isolating bidirectional controls" means either markup or formatting characters, this should probably be mentioned.

aphillips replied:
Done.

aphillips committed:
Address @r12a's comment about language tags.

This resulted in a bit of a rewrite.

In addition, I added links to the "choosing a language tag" article,
Richard's language subtag lookup tool, and the registry itself.

aphillips committed:
Address @r12a's comment about "isolating bidi controls"

aphillips committed:
Fixed bidi example by using markup.

A couple of browsers don't seem to display the "fixed" bidi example
correctly, at least when respec runs on it. Switching to markup, which
is effective.

Also rewrite the "notice how" paragraph to be (hopefully) clearer.

aphillips committed:
Address teleconference comments.

aphillips committed:
Improve color contrast in example 3

aphillips committed:
Further simplified bidi example.
- made the example text larger to make the color mixing clearer
- simplified the description of the mixing problem

aphillips committed:
Additional work on sorting section.
- Removed the charmod reqs altogether
- Added a "warning" box about find being hard and linking our
  string-search doc

xfq commented:
I noticed that Netlify wasn't working. It seems like someone revoked access to all W3C repos (not just the i18n ones). I'll try to add it back.

aphillips committed:
Fix description of BCP47 and of matching
- Removed extreneous link to specdev?
- Described matching in terms of ranges rather than subtags (4647 is not
  about matching subtags)
- Cleaned up description of 5656, including mentioning the LSTR

aphillips committed:
Minor tweak to ascending code unit sort.

Made the example clear that the code units are in the UTF-16 character
encoding.

aphillips committed:
Added text to description of the matching BP

aphillips committed:
Address comments on #bidi_gen section from telecon of 2023-02-09
- Added a requirement not to concat strings!
- Added a "seealso" suggested by @r12a
- Added explanation of the Note example
- Minor wording changes in the colorful example.

aphillips committed:
Move explanation under mustard.
The example note should be preceeded by the explanation.

aphillips committed:
Ensure all parts of 639 are dealt with, minor edits

aphillips committed:
Further rephrasing based on @r12a's comments in telecon

aphillips committed:
Further edits to make reqs refer to string values.

aphillips committed:
Address telecon 2023-02-16 comments
- add exampleChar style (borrowed from our other specs)
- renamed bidi_gen to strcat
- clarified further the concat section
- added links to uts10 and icu user guide for sorting
- moved langtag matching to its own section and fixed label

xfq commented on local.css:
```diff
@@ -431,3 +431,9 @@ table.whitespace caption {
+td.exampleChar {
+  font-size: 36pt;
+  text-align: center;
+  vertical-align: center;
```
This should probably be `vertical-align: middle;`. See https://w3c.github.io/csswg-drafts/css2/#propdef-vertical-align

aphillips replied:
thanks. Fixed in 07cbc60.

aphillips committed:
Address @xfq's comment about vertical-align

aphillips committed:
Update bidi recommendation from telecon of 2023-03-09 [I18N-ACTION-1253]

aphillips commented:
@r12a Could you look at this for merging-readiness? I'd like to include these changes when pushing to TR...

r12a commented on index.html:
```diff
@@ -937,9 +947,58 @@ <h3><em>Detecting &amp; matching direction (TBD)</em></h3>
 
+<p class="reviewComments"><a href="https://github.com/w3c/i18n-activity/labels/t%3Astrcat" target="_blank">See related review comments.</a></p>
+
+<div class="req" id="strcat_anti_pattern">
```
I don't think people will find this general advice about composite messages, that has nothing to do with text direction.

I think it is best to move the section "Composite Strings" under 'Text-processing" and then add a 'see also' to the text direction section 3.6.

aphillips replied:
pending telecon discuss

r12a commented on index.html:
```diff
@@ -937,9 +947,58 @@ <h3><em>Detecting &amp; matching direction (TBD)</em></h3>
+</div>
+
+<div class="req" id="bidi_gen_advice">
+	<p class="advisement">Specifications SHOULD provide guidance to implementers on avoiding problems with text directionality for any field(s) that the specification requires an implementation to create or generate for display to users.</p>
```
Rather a complicated and long sentence.

How about:
When a specification requires an implementation to create or generate text which will be displayed to users, the specification should provide implementers with guidance on
how to avoid potential problems related to text direction.

aphillips replied:
Done.

r12a commented on index.html:
```diff
@@ -937,9 +947,58 @@ <h3><em>Detecting &amp; matching direction (TBD)</em></h3>
+	<p class="advisement">Specifications SHOULD provide guidance to implementers on avoiding problems with text directionality for any field(s) that the specification requires an implementation to create or generate for display to users.</p>
+</div>
+
+<p>Specifications for APIs, protocols, or document formats sometimes require an implementation to create or provide a field containing a display name or description. When such a string is assembled from separate parts, it can result in problems with presentation or understanding due to the way that the <cite>Unicode Bidirectional Algorithm</cite> [[UAX9]] processes the assembled string. In such c ...
```
Several things are unclear to me here, and rather than provide text suggestions that may be barking up the wrong tree, we should perhaps discuss during our telecon.  Questions include: 
- do you mean bidrectional text or text that generally runs in a particular direction (eg. RTL)?
- i suspect, but i'm not sure, that we are talking solely about strings that contain bidirectional text within them, or about correctly displaying any string that is inserted into a target location by the consumer?
- if the latter, we're focusing on what producers need to do, we should mention dirname or using other metadata

Basically, i think we need a little more clarity about the role of producers vs consumers, and in-string direction vs how a string will be inserted into a target location.  It seems a little vague at the moment.  The section title, Concatenation of Strings, doesn't make it sound like we're concerned with string internal direction, but rather strings being inserted into the target location.  The example shows one scenario that fits the section title, but the preceding text needs to be clearer before we get to the example, and needs to clarify whether we are also talking about string-internals.

aphillips replied:
added to the agenda for 2023-03-30

r12a commented on index.html:
```diff
@@ -2779,42 +2838,103 @@ <h3>Specifying sort and search functionality</h3>
 
-
-    <p>Applications often need to organize sets of information or content. Frequently this involves sorting the content so that users can find what they are looking for. Many data types, such as numbers or dates, are easily sorted by comparing the values. When it comes to textual information, however, the nature of character encodings brings some additional complexity.</p>
+    <p>Applications often need to organize sets of information or content. Frequently this involves sorting the content. Many data types, such as numbers or dates, are easily sorted by comparing the values. When it comes to textual information, however, the nature of character encodings and user expectations regarding "alphabetical" order brings some additional complexity.</p>
```
Numbers and dates are only easy to sort if you compare internal representations of the data.

aphillips replied:
Done with:

```suggestion
Many non-textual data types, such as numbers or dates, can be easily sorted using the internal data representation. 
```

r12a commented on index.html:
```diff
@@ -2828,6 +2948,12 @@ <h3>Specifying sort and search functionality</h3>
 	</div>
+-->
+
+	<aside class="warning" id="find-is-hard" title="String search is hard">
```
As a warning about _string searching_ this seems a little oddly placed here.  I think that similar difficulties apply to matching strings for sorting, but i think it would be useful to couch this warning in those terms.

aphillips replied:
This section is (and has historically been) called "specifying sort and search functionality". All of the prose in the section is about _sorting_. This is the only mention of "search"--I added it so that we wouldn't have a gap. I think maybe a subsection div above it would help. I also retitled as "under construction". See the new PR

r12a commented on local.css:
```diff
@@ -431,3 +431,9 @@ table.whitespace caption {
 }
+
+td.exampleChar {
+  font-size: 36pt;
```
Can we use a rem or, if necessary, an em unit rather than pt?  If you really want to fix the size, perhaps use px rather than pt.

aphillips replied:
Done. Made it `300%`
